compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -1201,6 +1203,71 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
+    fn suggest_range_for_iter(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        actual: Ty<'tcx>,
+        source: SelfSource<'tcx>,
+        span: Span,
+        item_name: Ident,
+        ty_str: &str,
+    ) -> bool {
+        if let SelfSource::MethodCall(expr) = source {
+            for (_, parent) in tcx.hir().parent_iter(expr.hir_id) {
+                if let Node::Expr(parent_expr) = parent && is_range_literal(parent_expr) {
+                    let span_included = match parent_expr.kind {
+                            hir::ExprKind::Struct(_, eps, _) => eps.iter().any(|ep| ep.expr.span.contains(span)),
+                            // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
+                            hir::ExprKind::Call(ref func, ..) => func.span.contains(span),
+                            _ => false,
+                    };
+
+                    if !span_included {
+                        continue;
+                    }
+
+                    let range_def_id = self.tcx.lang_items().range_struct().unwrap();
+                    let range_ty = self.tcx.bound_type_of(range_def_id).subst(self.tcx, &[actual.into()]);
+                    let call_expr = self.tcx.hir().expect_expr(self.tcx.hir().get_parent_node(expr.hir_id));
+
+                    // avoid suggesting when the method name is not implemented for a `range`
+                    let pick =  self.lookup_probe(
+                        span,
+                        item_name,
+                        range_ty,
+                        call_expr,
+                        ProbeScope::AllTraits
+                    );
+
+                    if pick.is_ok() {
+                        let mut err = struct_span_err!(
```

Comment:
Can you make this into a structured diagnostic? https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-structs.html